### ISSUE TITLE
Strapi admin console for prod installation

### ISSUE THREAD

**Issue by jmchza** [strapi/strapi#1993]
**Informations**
- **Node.js version**:10.10.0
- **npm version**:6.4.1
- **Strapi version**:3.0.0-alpha.14.1.1
- **Database**:Postgres
- **Operating system**: docker container , image: Node:10-alpine 

**What is the current behavior?**
After the installation for a production version the next messages occur on npm start
ERROR in Entry module not found: Error: Can't resolve '/Workspaces/Storyx-WKS/my-forks-ones/strapi-plugin-dev/test-project/plugins/content-manager/node_modules/strapi-helper-plugin/lib/src/app.js' in '/Workspaces/Storyx-WKS/my-forks-ones/strapi-plugin-dev/test-project/admin'

ERROR in Entry module not found: Error: Can't resolve '/Workspaces/Storyx-WKS/my-forks-ones/strapi-plugin-dev/test-project/plugins/content-type-builder/node_modules/strapi-helper-plugin/lib/src/app.js' in '/Workspaces/Storyx-WKS/my-forks-ones/strapi-plugin-dev/test-project/admin'

ERROR in Entry module not found: Error: Can't resolve '/Workspaces/Storyx-WKS/my-forks-ones/strapi-plugin-dev/test-project/plugins/email/node_modules/strapi-helper-plugin/lib/src/app.js' in '/Workspaces/Storyx-WKS/my-forks-ones/strapi-plugin-dev/test-project/admin'

ERROR in Entry module not found: Error: Can't resolve '/Workspaces/Storyx-WKS/my-forks-ones/strapi-plugin-dev/test-project/plugins/settings-manager/node_modules/strapi-helper-plugin/lib/src/app.js' in '/Workspaces/Storyx-WKS/my-forks-ones/strapi-plugin-dev/test-project/admin'

ERROR in Entry module not found: Error: Can't resolve '/Workspaces/Storyx-WKS/my-forks-ones/strapi-plugin-dev/test-project/plugins/upload/node_modules/strapi-helper-plugin/lib/src/app.js' in '/Workspaces/Storyx-WKS/my-forks-ones/strapi-plugin-dev/test-project/admin'

ERROR in Entry module not found: Error: Can't resolve '/Workspaces/Storyx-WKS/my-forks-ones/strapi-plugin-dev/test-project/plugins/users-permissions/node_modules/strapi-helper-plugin/lib/src/app.js' in '/Workspaces/Storyx-WKS/my-forks-ones/strapi-plugin-dev/test-project/admin'

**Steps to reproduce the problem**
create a new project as the contribution guide specifies it and run the setup command.
After that npm start throws the above.

**What is the expected behavior?**
same as in development but with no admin directory linked to the strapi installation

**Suggested solutions**
I need help so please point out who had solve this, I am sure many ppl has already bump into this :(

**Comment by jmchza** [comment]
After copied the stra-helper-plugin from the strapi installation or original repository into the plugins directory, it throws the following error:

> strapi-admin@3.0.0-alpha.14.1.1 prestart /Workspaces/Storyx-WKS/my-forks-ones/strapi-plugin-dev/test-project/admin
> node ./node_modules/strapi-helper-plugin/node_modules/.bin/cross-env NODE_ENV=development PORT=4000 IS_ADMIN=true node ./node_modules/strapi-helper-plugin/lib/internals/scripts/loadAdminConfigurations.js

(node:49028) UnhandledPromiseRejectionWarning: TypeError: Cannot read property 'routes' of undefined
    at _.forEach (/Workspaces/Storyx-WKS/my-forks-ones/strapi-plugin-dev/test-project/plugins/users-permissions/middlewares/users-permissions/index.js:31:35)
    at /Workspaces/Storyx-WKS/my-forks-ones/strapi-plugin-dev/test-project/plugins/users-permissions/node_modules/lodash/lodash.js:4911:15
    at baseForOwn (/Workspaces/Storyx-WKS/my-forks-ones/strapi-plugin-dev/test-project/plugins/users-permissions/node_modules/lodash/lodash.js:2996:24)
    at /Workspaces/Storyx-WKS/my-forks-ones/strapi-plugin-dev/test-project/plugins/users-permissions/node_modules/lodash/lodash.js:4880:18
    at Function.forEach (/Workspaces/Storyx-WKS/my-forks-ones/strapi-plugin-dev/test-project/plugins/users-permissions/node_modules/lodash/lodash.js:9344:14)
    at Object.initialize (/Workspaces/Storyx-WKS/my-forks-ones/strapi-plugin-dev/test-project/plugins/users-permissions/middlewares/users-permissions/index.js:30:11)
    at /Workspaces/Storyx-WKS/my-forks-ones/strapi/packages/strapi/lib/middlewares/index.js:30:23
    at Promise (/Workspaces/Storyx-WKS/my-forks-ones/strapi/packages/strapi/lib/middlewares/index.js:130:43)
    at new Promise (<anonymous>)
    at Promise.all.Object.keys.map.middleware (/Workspaces/Storyx-WKS/my-forks-ones/strapi/packages/strapi/lib/middlewares/index.js:70:9)
    at Array.map (<anonymous>)
    at Strapi.module.exports (/Workspaces/Storyx-WKS/my-forks-ones/strapi/packages/strapi/lib/middlewares/index.js:68:34)
(node:49028) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). (rejection id: 3)
(node:49028) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.

What we suppose to do here? The contributing guide seems to be done for development of plugins not for production.
thanks in advance.

**Comment by soupette** [comment]
@jmchza what do you mean installation for production version? You did some modifications into a plugin and you want them to be in your app that is not generated with `--dev` ?

**Comment by jmchza** [comment]
OK I think I found out the confusion.

for prod deploy with the default plugins we need to run the build:setup on the strapi installation.
that was the root cause of all those above issues.

after that awe can proceed with the following steps:
1- strapi new my-project
2 - cd ./my-project
3 - npm install && npm postinstall
4 - cd admin
5 - npm install
6- cd node_modules/strapi-helper-plugin
7- npm install

 and from the admin folder we run the npm start

I have followed this steps and it finally did work. but the guide does not say anything about the helper-plugin compilation :( 

For everyone having trouble to generate a prod version of the admin console, I believe those stepp generate you a fresh one and working ok.

**Comment by soupette** [comment]
The other solution would be the following

```
# create a new app in dev mode
$ strapi new my-app --dev
# when you're done with your modifications
$ cd my-app && npm run setup --plugins
# create a prod app 
$ strapi new my-prod-app
# then just copy the build folder into the corresponding plugins
# check that the build is ok and you have no babel-polyfill error
$ cd my-prod-app && strapi-start
```

**Comment by nickerlan** [comment]
@soupette, when I used this solution I've got: 

📦  Installing packages...
Packaged installed successfully

> strapi-admin@3.0.0-alpha.14.1.1 setup /usr/local/lib/node_modules/strapi/node_modules/strapi-admin
> node ./scripts/setup.js

🏗  Building the admin...
internal/modules/cjs/loader.js:550
    throw err;
    ^

Error: Cannot find module '/usr/local/lib/node_modules/strapi/node_modules/strapi-admin/node_modules/strapi-helper-plugin/lib/internals/scripts/loadAdminConfigurations.js'

**Comment by jmchza** [comment]
@soupette Thanks for you response.
What about running the admin console pointing to a separate api on another server ? 
that would be running the admin console with the addition of this, but not DB associated locally:

I would really appreciate your clarification on this one. thank you heaps in advance.

```
"admin": {
    "build": {
      "host": "http://localhost:4000/admin",
      "backend": "https://mydomain.com",
      "plugins": {
        "source":  "host",
        "folder": "/plugins"
      }
    }
```

would this work or the admin console would still to have access to the DB in order to start properly?
I just tried it and I get this errors:
```
[2018-09-21T12:49:10.128Z] info strapi-local-auth initialized
(node:2245) UnhandledPromiseRejectionWarning: Error: connect ECONNREFUSED 127.0.0.1:5432
    at TCPConnectWrap.afterConnect [as oncomplete] (net.js:1113:14)
From previous event:
    at Client_PG.acquireRawConnection (/Workspaces/Storyx-WKS/admin-new/node_modules/knex/lib/dialects/postgres/index.js:147:12)
    at create (/Workspaces/Storyx-WKS/admin-new/node_modules/knex/lib/client.js:280:23)
    at tryPromise (/Workspaces/Storyx-WKS/admin-new/node_modules/tarn/lib/Pool.js:366:22)
    at tryPromise (/Workspaces/Storyx-WKS/admin-new/node_modules/tarn/lib/utils.js:57:20)
    at Promise (/Workspaces/Storyx-WKS/admin-new/node_modules/tarn/lib/Pool.js:366:5)
    at new Promise (<anonymous>)
    at callbackOrPromise (/Workspaces/Storyx-WKS/admin-new/node_modules/tarn/lib/Pool.js:357:10)
    at Pool._create (/Workspaces/Storyx-WKS/admin-new/node_modules/tarn/lib/Pool.js:307:5)
    at Pool._doCreate (/Workspaces/Storyx-WKS/admin-new/node_modules/tarn/lib/Pool.js:275:32)
    at Pool._tryAcquireOrCreate (/Workspaces/Storyx-WKS/admin-new/node_modules/tarn/lib/Pool.js:212:12)
    at Pool.acquire (/Workspaces/Storyx-WKS/admin-new/node_modules/tarn/lib/Pool.js:119:10)
    at /Workspaces/Storyx-WKS/admin-new/node_modules/knex/lib/client.js:335:26
From previous event:
    at Client_PG.acquireConnection (/Workspaces/Storyx-WKS/admin-new/node_modules/knex/lib/client.js:334:34)
    at Runner.ensureConnection (/Workspaces/Storyx-WKS/admin-new/node_modules/knex/lib/runner.js:228:24)
    at Runner.run (/Workspaces/Storyx-WKS/admin-new/node_modules/knex/lib/runner.js:34:42)
    at SchemaBuilder.Target.then (/Workspaces/Storyx-WKS/admin-new/node_modules/knex/lib/interface.js:20:43)
(node:2245) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). (rejection id: 8)
(node:2245) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.
(node:2245) UnhandledPromiseRejectionWarning: Error: connect ECONNREFUSED 127.0.0.1:5432
    at TCPConnectWrap.afterConnect [as oncomplete] (net.js:1113:14)
```

**Comment by jmchza** [comment]
@soupette I am only trying to deploy a prod version with no modifications, therefore i only run this:
$ strapi new my-prod-app
# then just copy the build folder into the corresponding plugins
# check that the build is ok and you have no babel-polyfill error
$ cd my-prod-app && strapi-start
and put that content into a repository that then gets deployed.
but I am having issues to run it as for some reason the strapi-plugin on the strapi-admin@3.0.0-alpha.14.1.1 version seems to have some issues or missing linkage.

this component always misses and I have tries so many things already and none of them worked. I've been spending 30 hours doing this and still no success.

is it possible that on that version the npm install trys to use the 3@0.1.0 node node_modules/strapi/lib/utils/post-install.js instead of the 3.0.0-alpha.14.1.1  and therefore fails ?

```
root@c7126b4b75b9:/opt/faktoora# npm install
npm WARN lifecycle faktoora-strapi-v3@0.1.0~postinstall: cannot run in wd faktoora-strapi-v3@0.1.0 node node_modules/strapi/lib/utils/post-install.js (wd=/opt/faktoora)
npm WARN bookshelf@0.12.1 requires a peer of knex@>=0.6.10 <0.15.0 but none is installed. You must install peer dependencies yourself.
npm WARN bootstrap@4.1.3 requires a peer of jquery@1.9.1 - 3 but none is installed. You must install peer dependencies yourself.
npm WARN ajv-keywords@3.2.0 requires a peer of ajv@^6.0.0 but none is installed. You must install peer dependencies yourself.
npm WARN eslint-config-airbnb@15.1.0 requires a peer of eslint-plugin-jsx-a11y@^5.1.1 but none is installed. You must install peer dependencies yourself.
npm WARN optional SKIPPING OPTIONAL DEPENDENCY: fsevents@1.2.4 (node_modules/fsevents):
npm WARN notsup SKIPPING OPTIONAL DEPENDENCY: Unsupported platform for fsevents@1.2.4: wanted {"os":"darwin","arch":"any"} (current: {"os":"linux","arch":"x64"})

audited 31062 packages in 18.044s
```

I will really appreciate some insight from you as it is hard to take assumptions without getting into a deep whole :( 

Thank you in advance.

**Comment by jmchza** [comment]
for some reason after the generated one gets placed into a repository it does not compile in the container as expected:

After running the postinstall script in my local env i get a lot of component in the node_modules but in the container for some reason U only get these ones, but the most important one does not get created: strapi-helper-plugin

There is a huge missing point here at the time a prod version is intending to be deployed. guys these are the things that can put huge amount of ppl away of this. 

By a prod version I mean only a standard one with no plugins modifications and ready to be used as a prod.

What is the recommended way here: do we need to downgrade the npm version? if so please do point out one tha works with this strapi alpha version.

Much appreciated in advance.

```
bash-4.4# ll admin/node_modules/
total 108
drwxr-xr-x   27 root     root          4096 Sep 22 21:53 .
drwxrwxrwx    1 root     root          4096 Sep 22 21:53 ..
drwxr-xr-x    2 root     root          4096 Sep 22 21:53 .bin
drwxr-xr-x    2 root     root          4096 Sep 22 21:53 balanced-match
drwxr-xr-x    2 root     root          4096 Sep 22 21:53 brace-expansion
drwxr-xr-x    4 root     root          4096 Sep 22 21:53 concat-map
drwxr-xr-x    2 root     root          4096 Sep 22 21:53 fs.realpath
drwxr-xr-x    2 root     root          4096 Sep 22 21:53 glob
drwxr-xr-x    2 root     root          4096 Sep 22 21:53 inflight
drwxr-xr-x    2 root     root          4096 Sep 22 21:53 inherits
drwxr-xr-x    2 root     root          4096 Sep 22 21:53 interpret
drwxr-xr-x    2 root     root          4096 Sep 22 21:53 js-tokens
drwxr-xr-x    2 root     root          4096 Sep 22 21:53 loose-envify
drwxr-xr-x    2 root     root          4096 Sep 22 21:53 minimatch
drwxr-xr-x    2 root     root          4096 Sep 22 21:53 object-assign
drwxr-xr-x    2 root     root          4096 Sep 22 21:53 once
drwxr-xr-x    2 root     root          4096 Sep 22 21:53 path-is-absolute
drwxr-xr-x    2 root     root          4096 Sep 22 21:53 path-parse
drwxr-xr-x    3 root     root          4096 Sep 22 21:53 prop-types
drwxr-xr-x    4 root     root          4096 Sep 22 21:53 react
drwxr-xr-x    7 root     root          4096 Sep 22 21:53 react-ga
drwxr-xr-x    3 root     root          4096 Sep 22 21:53 rechoir
drwxr-xr-x    3 root     root          4096 Sep 22 21:53 remove-markdown
drwxr-xr-x    5 root     root          4096 Sep 22 21:53 resolve
drwxr-xr-x    4 root     root          4096 Sep 22 21:53 schedule
drwxr-xr-x    4 root     root          4096 Sep 22 21:53 shelljs
drwxr-xr-x    2 root     root          4096 Sep 22 21:53 wrappy
```

**Comment by soupette** [comment]
@nickerlan can you check that Strapi is installed correctly and that you have the missing file in your node_modules?
To do so open a terminal window run `npm root -g` copy the path and run `open .`. In the strapi folder you should see the `strapi-admin`. Can you make sure the `strapi-helper-plugin` is there?
(I've never encountered this error..).

@jmchza after following my steps did you get the server running and also the admin available?

**Comment by jmchza** [comment]
Yes those steps were verified @soupette, thanks for that.

I will close this now, I have moved my priorities since I have spent too much time on this.
Cheers,
M

**Comment by mdocter** [comment]
I couldn't get the admin console to work in production. Same errors as mentioned above.
And 404 page and 'No plugins installed yet.' messages. It was driving me crazy.
But I just found the cause of my issues... I followed the bad advice:

`We highly recommend to change the default `/admin` path for security reasons.`

Like mentioned in the following doc: https://strapi.io/documentation/3.x.x/guides/deployment.html#deployment

Changing it back to `/admin` fixed it.
